File: frontend_admin/lib/api/update.dart
import 'dart:convert';
import 'package:dio/dio.dart';
import 'package:http/http.dart' as http;
import 'package:nimbus/api/constants.dart';
 
class UpdateProjectRequest {
  final String? name;
  final String? description;
  final int? year;
  final String? country;
  final List<String>? pictures;
  final List<String>? videos;

  UpdateProjectRequest({
    this.name,
    this.description,
    this.year,
    this.country,
    this.pictures,
    this.videos,
  });

  Map<String, dynamic> toJson() => {
        'name': name,
        'description': description,
        'year': year,
        'country': country,
        'pictures': pictures,
        'videos': videos,
      };
}

Future<bool> updateProject(int projectId, UpdateProjectRequest project) async {
  final url = "$baseUrl/api/projects/$projectId";
  try {
    final response = await Dio().patch(url, data: project.toJson());
    if (response.statusCode == 200) {
      return true;
    } else {
      print('Failed to update project: ${response.statusCode}');
      return false;
    }
  } catch (e) {
    print("Cought: $e");
    return false;
  }

}

File: backend/src/admin/projects/update.rs
use crate::{entities, state};
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::{IntoResponse, Response},
    Json,
};
use entities::projects;
use sea_orm::{ActiveModelTrait, DbErr, EntityTrait};
use serde::{Deserialize, Serialize};
use state::AppState;

#[derive(thiserror::Error, Debug)]
pub enum UpdateProjectError {
    #[error("No project id({0}) found")]
    NoProjectFound(i32),

    #[error("Db error: {0}")]
    DbError(#[from] DbErr),
}

impl IntoResponse for UpdateProjectError {
    fn into_response(self) -> Response {
        (
            StatusCode::INTERNAL_SERVER_ERROR,
            format!("There was a problem: {}", self)
        )
            .into_response()
    }
}

#[derive(Deserialize, Serialize, Default)]
pub struct UpdateProjectResponse {
    id: i32,
}

#[derive(Deserialize, Debug)]
pub struct UpdateProjectRequest {
    pub name: Option<String>,
    pub description: Option<String>,
    pub year: Option<i32>,
    pub country: Option<String>,

    pub pictures: Option<Vec<String>>,
    pub videos: Option<Vec<String>>,
}

pub async fn project(
    State(state): State<AppState>,
    Path(project_id): Path<i32>,
    Json(info): Json<UpdateProjectRequest>
) -> Result<StatusCode, UpdateProjectError> {
    let existing_project = projects::Entity::find_by_id(project_id)
        .one(&state.db_conn)
        .await?
        .ok_or(UpdateProjectError::NoProjectFound(project_id))?;

    let mut project: projects::ActiveModel = existing_project.into();

    if let Some(name) = info.name {
        project.name = sea_orm::Set(name);
    }

    if let Some(description) = info.description {
        project.description = sea_orm::Set(description);
    }

    if let Some(year) = info.year {
        project.year = sea_orm::Set(year);
    }

    if let Some(country) = info.country {
        project.country = sea_orm::Set(country);
    }

    if let Some(pictures) = info.pictures {
        project.pictures = sea_orm::Set(pictures);
    }

    if let Some(videos) = info.videos {
        project.videos = sea_orm::Set(videos);
    }

    let _res = project.update(&state.db_conn).await?;

    Ok(StatusCode::OK)
}
